caffe2/core/dispatch/OpSchema.h

```diff
@@ -0,0 +1,233 @@
+    static_assert(std::is_same<
+      map_t<guts::remove_cv_t, map_t<guts::remove_reference_t, typelist<Args...>>>,
+      map_t<guts::remove_cv_t, map_t<guts::remove_reference_t, typename signature::parameter_types>>
+      >::value, "Invalid argument types passed to OpSchema::dispatch_key()");
+    return dispatch_key_type {
+      details::getTensorTypeIds_(args...)
+    };
+  }
+};
+
+// Special case. Operator overwrites DispatchKey generation. Use that.
+template<class OpSchemaDef>
+class OpDispatchKeySchema<OpSchemaDef, guts::enable_if_t<has_function_dispatch_key_defined<OpSchemaDef>::value>> final {
+  using signature = OpSignatureSchema<OpSchemaDef>;
+
+  static_assert(guts::is_function_type<decltype(OpSchemaDef::dispatch_key)>::value, "Operator schema defines dispatch_key member, but it isn't a function.");
+
+  using dispatch_key_traits = guts::function_traits<decltype(OpSchemaDef::dispatch_key)>;
+
+public:
+  using dispatch_key_type = typename dispatch_key_traits::return_type;
+
+private:
+
+  static_assert(guts::is_equality_comparable<dispatch_key_type>::value, "Operator schema specified custom dispatch_key() derivation function, but the returned dispatch key type doesn't have the equality operator defined. Please define it.");
+  static_assert(guts::is_hashable<dispatch_key_type>::value, "Operator schema specified custom dispatch_key() derivation function, but the returned dispatch key type doesn't have an overload for std::hash. Please define it.");
+
+  static_assert(std::is_same<
+    guts::typelist::map_t<guts::remove_cv_t, guts::typelist::map_t<guts::remove_reference_t, typename dispatch_key_traits::parameter_types>>,
+    guts::typelist::map_t<guts::remove_cv_t, guts::typelist::map_t<guts::remove_reference_t, typename signature::parameter_types>>
+    >::value, "Operator schema defines custom dispatch_key() derivation function, but the arguments don't match the operator signature.");
+
+public:
+
+  template<class... Args>
+  static inline dispatch_key_type dispatch_key(const Args&... args) {
+    using guts::typelist::map_t;
+    using guts::typelist::typelist;
+    static_assert(std::is_same<
+      map_t<guts::remove_cv_t, map_t<guts::remove_reference_t, typelist<Args...>>>,
+      map_t<guts::remove_cv_t, map_t<guts::remove_reference_t, typename signature::parameter_types>>
+      >::value, "Invalid argument types passed to OpSchema::dispatch_key()");
+    return OpSchemaDef::dispatch_key(args...);
+  }
+};
+
+}  // namespace details
+
+/**
+ * Wrapper class for user-defined OpSchemaDef, providing functionality for determining
+ * information about the signature and dispatching on that signature.  This is the
+ * "public" facing class.
+ *
+ * @tparam OpSchemaDef User-defined OpSchemaDef.
+ *   This struct is expected to define:
+ *      - a function type Signature
+ *      - a constexpr guts<const char*, n_args> parameter_names field (where n_args is
+ *        the number of arguments in Signature)
```

Comment:
This API is likely to have UI changes coming down the pipe, right? Note it down here.

Reply:
What do you mean UI changes? Changes to how OpSchemaDef has to be written or changes to how OpSchema exposes the parsed structure?